### PR TITLE
Write Pact consumer tests for account-api 

### PR DESCRIPTION
**Pull request by barrucadu**
This is step 1 of the [Writing Pact tests](https://docs.publishing.service.gov.uk/manual/pact-broker.html#writing-pact-tests) docs, see also https://github.com/alphagov/account-api/pull/13.

---

Pact consumer tests have a few different parts:

1. They define request/response templates (much like webmock, but with
a different DSL); this is the `before` blocks

2. They run the tests and complain if:
   - any unexpected requests are made
   - any defined requests aren't made

The templates are saved to a JSON file, which is used by the provider
tests (in this case, the account-api tests) to validate that the
provider returns matching outputs for the same inputs.

The "given" part of the set-up corresponds to a "provider state",
which is a named `before` block, run as part of the provider tests.

---

[Trello card](https://trello.com/c/OcRffYwv/674-set-up-pact-tests-for-our-account-api-app)

### CODE REVIEW

File: test/account_api_test.rb [hunk 1]
```diff
@@ -0,0 +1,407 @@
...
+          method: :get,
+          path: "/api/oauth2/sign-in",
+          headers: GdsApi::JsonClient.default_request_headers,
+        )
+        .will_respond_with(
+          status: 200,
+          headers: { "Content-Type" => "application/json; charset=utf-8" },
+          body: {
+            auth_uri: Pact.like("string"),
+            state: Pact.like("string"),
```
Comment (ChrisBAshton) [inline test/account_api_test.rb:29]: I think [Pact matching convention](https://docs.pact.io/getting_started/matching/) is to put an example value here, rather than describing the type. Pact's own matching will infer that the example value is a string.

File: test/account_api_test.rb [hunk 2]
```diff
@@ -0,0 +1,407 @@
...
+            status: 200,
+            headers: { "Content-Type" => "application/json; charset=utf-8" },
+            body: {
+              govuk_account_session: Pact.like("string"),
+              has_subscription: has_subscription,
+            },
+          )
+      end
+
+      it "responds with 200 OK, a new govuk_account_session, and says that the subscription does not exists" do
```
Comment (ChrisBAshton) [inline test/account_api_test.rb:190]:
```suggestion
      it "responds with 200 OK, a new govuk_account_session, and says that the subscription does not exist" do
```

File: test/account_api_test.rb [hunk 3]
```diff
@@ -0,0 +1,407 @@
...
+        assert response["govuk_account_session"].present?
+        assert_equal false, response["has_subscription"]
+        assert_equal 200, response.code
+      end
+
+      describe "a subscription does not exist" do
+        let(:given) { "there is a valid user session, with a transition checker email subscription" }
+        let(:has_subscription) { true }
+
+        it "says that the subscription does not exist" do
```
Comment (ChrisBAshton) [inline test/account_api_test.rb:201]: Is that what's being asserted below? 🤔

**Review by ChrisBAshton — COMMENTED** [review]
Approach LGTM 👍 Just a couple of minor comments below.

File: test/account_api_test.rb [hunk 4]
```diff
@@ -1,81 +1,407 @@
...
+        .will_respond_with(
+          status: 422,
+          headers: { "Content-Type" => "application/json; charset=UTF-8" },
+        )
+    end
+
+    it "responds with 422 Unprocessable Entity" do
+      assert_raises GdsApi::HTTPUnprocessableEntity do
+        api_client.get_sign_in_url(redirect_path: redirect_path)
+      end
```
Comment (kevindew) [inline test/account_api_test.rb:64]: A quick observation of mine is that we probably don't need pact tests for negative scenarios when this app doesn't have any code to handle those scenarios. It doesn't seem that app needs a contract that Account API will act this way in a particular scenario because it doesn't have a view on it.

(I imagine this is a carryover from Publishing API ones that may have a bunch of scenarios tested that may not actually add value).

The positive ones I've glanced at look good though. I've not got time for a full review but awesome to see you're working on these and seeking feedback 👍

File: test/account_api_test.rb (hunk 3 above)
Comment (barrucadu) [reply test/account_api_test.rb:201]: Whoops, I flipped the logic of this one and didn't update the name

File: test/account_api_test.rb [hunk 5]
```diff
@@ -1,81 +1,407 @@
...
-      api_client.validate_auth_response(code: "foo", state: "bar")
+    before do
+      account_api
+        .given(given)
+        .upon_receiving("a validation request")
+        .with(
+          method: :post,
+          path: "/api/oauth2/callback",
+          body: {
+            code: Pact.like("code"),
```
Comment (hannako) [inline test/account_api_test.rb:80]: You could use `Pact.like(code: "code", state: "state")` as you have on line 121. Similar comment for line 153

Comment (barrucadu) [reply test/account_api_test.rb:80]: I think I'll actually just change those to a hash.  I've just been reading https://docs.pact.io/consumer/, which says

> Typically, exact matching is most appropriate for Pact tests on the consumer side that are running at the unit test level. The same person is responsible for both the expectation and the actual request, so making sure that they match should be straightforward.

File: test/account_api_test.rb (hunk 4 above)
Comment (barrucadu) [reply test/account_api_test.rb:64]: Sure, I see where you're coming from. And reading [the pact docs](https://docs.pact.io/consumer/), it says you shouldn't use pact tests to ensure functional correctness of the provider.  So I'll remove the error case tests.

File: test/account_api_test.rb [hunk 6]
```diff
@@ -1,81 +1,407 @@
...
+            state: Pact.like("state"),
+          },
+          headers: GdsApi::JsonClient.default_request_with_json_body_headers,
+        )
+        .will_respond_with(
+          status: 200,
+          headers: { "Content-Type" => "application/json; charset=utf-8" },
+          body: {
+            govuk_account_session: Pact.like("user-session-id"),
+            redirect_path: Pact.like(redirect_path),
```
Comment (hannako) [inline test/account_api_test.rb:64]: I found it a bit confusing that `redirect_path` is nil. What does Pact.like(nil) allow to pass?

Comment (barrucadu) [reply test/account_api_test.rb:64]: Ah, this can go away now.  It was useful when I had the test for the error case, so I didn't have to write two almost-identical before blocks, but now that's gone.

File: test/account_api/account_api_test.rb [hunk 7]
```diff
@@ -5,8 +5,7 @@
 describe GdsApi::AccountApi do
   include GdsApi::TestHelpers::AccountApi
 
-  let(:base_url)      { Plek.find("account-api") }
-  let(:api_client)    { GdsApi::AccountApi.new(base_url) }
+  let(:api_client) { GdsApi::AccountApi.new(Plek.find("account-api")) }
```
Comment (huwd) [inline test/account_api/account_api_test.rb:8]: Do we need the Plek line here?
I've been l[ooking through the GDS API](https://github.com/alphagov/gds-api-adapters/blob/603fb33f3894559029728eb9143d1644a7cd858e/lib/gds_api.rb#L32-L35)

Looked a bit like we could call `GdsApi.account_api` which also fetches stuff out of plek?

Comment (barrucadu) [reply test/account_api/account_api_test.rb:8]: Yeah, I did this to be consistent with the other tests

File: test/account_api_test.rb [hunk 8]
```diff
@@ -0,0 +1,278 @@
+require "test_helper"
+require "gds_api/account_api"
+
+describe GdsApi::AccountApi do
+  include PactTest
+
+  let(:api_client) { GdsApi::AccountApi.new(account_api_host) }
+
+  let(:authenticated_headers) { { "GOVUK-Account-Session" => govuk_account_session } }
```
Comment (huwd) [inline test/account_api_test.rb:9]: Given the travails in testing these over in frontend, might this be better as a standard capitalised + HTTP_ prefixed name?

Comment (huwd) [reply test/account_api_test.rb:9]:
```suggestion
  let(:authenticated_headers) { { "HTTP_GOVUK_ACCOUNT_SESSION" => govuk_account_session } }
```

File: test/account_api_test.rb [hunk 9]
```diff
@@ -0,0 +1,278 @@
...
+  let(:govuk_account_session) { "logged-in-user-session" }
```
Comment (huwd) [inline test/account_api_test.rb:10]: huh, I didn't know we could use let in minitest. I'd assumed that was an Rspec thing!
cool.

File: test/account_api_test.rb [hunk 10]
```diff
@@ -0,0 +1,278 @@
...
+
+  describe "getting a sign-in URL" do
+    let(:redirect_path) { nil }
+
+    before do
+      account_api
+        .upon_receiving("a sign-in request")
```
Comment (huwd) [inline test/account_api_test.rb:17]: These lines seem to just be human readable strings that don't repeat or match anywhere.
So what are they used for? Are these just Pact versions of `describe` blocks for test stack traces?

File: test/account_api_test.rb [hunk 11]
```diff
@@ -0,0 +1,278 @@
...
+      let(:given) { "there is a valid user session" }
+      let(:has_subscription) { false }
+
+      before do
+        account_api
+          .given(given)
+          .upon_receiving("a has-subscription request")
+          .with(
+            method: :get,
+            path: "/api/transition-checker-email-subscription",
```
Comment (huwd) [inline test/account_api_test.rb:143]: Now i see it alongside the others, this path in particular stands out like a sore thumb doesn't it.
So specific where the others are generic and flexible 🤔.

Makes me wonder about if we'll get a chance to tackle subscriptions integrations again and de-snowflake this.

File: test/test_helpers/pact_helper.rb [hunk 12]
```diff
@@ -1,6 +1,7 @@
 PUBLISHING_API_PORT = 3001
 ORGANISATION_API_PORT = 3002
 BANK_HOLIDAYS_API_PORT = 3003
+ACCOUNT_API_PORT = 3004
```
Comment (huwd) [inline test/test_helpers/pact_helper.rb:4]: Interesting that they all need to be on different ports 🤔 Why is that?
Is this as a precaution in case we ever run more than two of these in the same test block in future?

Or does it complain about a port clash even if we're not pact testing ACCOUNT_API and BANK_HOLIDAYS_API in the same request?

**Review by huwd — CHANGES_REQUESTED** [review]

File: test/account_api_test.rb (hunk 10 above)
Comment (barrucadu) [reply test/account_api_test.rb:17]: I think so, yes.  The `given`s correspond to set-up blocks in account-api, but these `upon_receiving`s are descriptive.

File: test/account_api_test.rb (hunk 8 above)
Comment (barrucadu) [reply test/account_api_test.rb:9]: No, the request header is `GOVUK-Account-Session` - the header-name-mangling is done by Rails: see https://github.com/alphagov/frontend/pull/2668#issuecomment-802843544

File: test/account_api_test.rb (hunk 11 above)
Comment (barrucadu) [reply test/account_api_test.rb:143]: Either we'll decide we don't need this feature once the transition checker goes away, and so we can drop it; or we'll decide we want it for more things, and so we'll have to rework it.  But I don't think there's much point in reworking it to be more generic until we need it.

File: test/test_helpers/pact_helper.rb (hunk 12 above)
Comment (barrucadu) [reply test/test_helpers/pact_helper.rb:4]: Pact throws an error if multiple services use the same port.  We might be able to fix that by moving the `Pact.service_consumer`s into the test files, but I think that's out of scope of the PR (and may be more complicated if we have to explicitly unregister the `service_consumer`s too).

File: test/account_api_test.rb (hunk 8 above)
Comment (huwd) [reply test/account_api_test.rb:9]: ah ha, great - cheers for the link!

**Review by huwd — APPROVED** [review]
